apis/client/v1beta1/openstackclient_webhook.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//
+// Generated by:
+//
+// operator-sdk create webhook --group client --version v1beta1 --kind OpenStackClient --programmatic-validation --defaulting
+//
+
+package v1beta1
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// OpenStackClientDefaults -
+type OpenStackClientDefaults struct {
+	ContainerImageURL string
+}
+
+var openStackClientDefaults OpenStackClientDefaults
+
+// log is for logging in this package.
+var openstackclientlog = logf.Log.WithName("openstackclient-resource")
+
+// SetupOpenStackClientDefaults - initialize OpenStackClient spec defaults for use with either internal or external webhooks
+func SetupOpenStackClientDefaults(defaults OpenStackClientDefaults) {
+	openStackClientDefaults = defaults
+	openstackclientlog.Info("OpenStackClient defaults initialized", "defaults", defaults)
+}
+
+// SetupWebhookWithManager sets up the webhook with the Manager
+func (r *OpenStackClient) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-client-openstack-org-v1beta1-openstackclient,mutating=true,failurePolicy=fail,sideEffects=None,groups=client.openstack.org,resources=openstackclients,verbs=create;update,versions=v1beta1,name=mopenstackclient.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Defaulter = &OpenStackClient{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *OpenStackClient) Default() {
+	openstackclientlog.Info("default", "name", r.Name)
+
+	r.Spec.Default()
+}
+
+// Default - set defaults for this OpenStackClient spec
+func (spec *OpenStackClientSpec) Default() {
+	if spec.ContainerImage == "" {
+		spec.ContainerImage = openStackClientDefaults.ContainerImageURL
+	}
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
```

Comment:
Is this comment still required?

Reply:
Across the operators I have been leaving these `TODO` comments from the webhook scaffolding in-place because they pertain to the validation webhooks, which are now present but haven't been utilized yet (this PR only adds actual logic for  the mutating/defaulting webhook, but see here [1] why both mutating and validating scaffolding are added at the same time, even if we're only using one type for the moment).  I can remove the comment, but it might be useful information for anyone who is looking to actually add validation logic.

[1] https://github.com/openstack-k8s-operators/docs/blob/main/webhooks.md#adding-webhooks-to-an-operator

Reply:
Clear, thanks for the pointer